src/openrct2/scripting/ScriptEngine.cpp

```diff
@@ -0,0 +1,956 @@
+        }
+        return DukToGameActionResult(dukResult);
+    }
+    else
+    {
+        auto action = std::make_unique<GameActionResult>();
+        action->Error = GA_ERROR::UNKNOWN;
+        action->ErrorTitle = "Unknown custom action";
+        return action;
+    }
+}
+
+std::unique_ptr<GameActionResult> ScriptEngine::DukToGameActionResult(const DukValue& d)
+{
+    auto result = std::make_unique<GameActionResult>();
+    result->Error = static_cast<GA_ERROR>(AsOrDefault<int32_t>(d["error"]));
+    result->ErrorTitle = AsOrDefault<std::string>(d["errorTitle"]);
+    result->ErrorMessage = AsOrDefault<std::string>(d["errorMessage"]);
+    result->Cost = AsOrDefault<int32_t>(d["cost"]);
+    return result;
+}
+
+DukValue ScriptEngine::PositionToDuk(const CoordsXYZ& position)
+{
+    DukStackFrame frame(_context);
+    duk_context* ctx = _context;
+    auto obj = duk_push_object(ctx);
+    duk_push_int(ctx, position.x);
+    duk_put_prop_string(ctx, obj, "x");
+    duk_push_int(ctx, position.y);
+    duk_put_prop_string(ctx, obj, "y");
+    duk_push_int(ctx, position.z);
+    duk_put_prop_string(ctx, obj, "z");
+    return DukValue::take_from_stack(ctx);
+}
+
+DukValue ScriptEngine::GameActionResultToDuk(const GameAction& action, const std::unique_ptr<GameActionResult>& result)
+{
+    DukStackFrame frame(_context);
+    DukObject obj(_context);
+
+    auto player = action.GetPlayer();
+    if (player != -1)
+    {
+        obj.Set("player", action.GetPlayer());
+    }
+    if (result->Cost != MONEY32_UNDEFINED)
+    {
+        obj.Set("cost", result->Cost);
+    }
+    if (!result->Position.isNull())
+    {
+        obj.Set("position", PositionToDuk(result->Position));
+    }
+
+    if (action.GetType() == GAME_COMMAND_CREATE_RIDE)
+    {
+        auto& rideCreateResult = static_cast<RideCreateGameActionResult&>(*result.get());
+        if (rideCreateResult.rideIndex != RIDE_ID_NULL)
+        {
+            obj.Set("ride", rideCreateResult.rideIndex);
+        }
+    }
+
+    return obj.Take();
+}
+
+bool ScriptEngine::RegisterCustomAction(
+    const std::shared_ptr<Plugin>& plugin, const std::string_view& action, const DukValue& query, const DukValue& execute)
+{
+    std::string actionz = std::string(action);
+    if (_customActions.find(actionz) != _customActions.end())
+    {
+        return false;
+    }
+
+    CustomAction customAction;
+    customAction.Owner = plugin;
+    customAction.Name = std::move(actionz);
+    customAction.Query = query;
+    customAction.Execute = execute;
+    _customActions[customAction.Name] = std::move(customAction);
+    return true;
+}
+
+void ScriptEngine::RemoveCustomGameActions(const std::shared_ptr<Plugin>& plugin)
+{
+    for (auto it = _customActions.begin(); it != _customActions.end();)
+    {
+        if (it->second.Owner == plugin)
+        {
+            it = _customActions.erase(it);
+        }
+        else
+        {
+            it++;
+        }
+    }
```

Comment:
Opportunity for the erase/remove_if idiom here too.
```suggestion
    auto isOwner = [&] (auto& obj) { return obj.second.Owner == plugin; }
    std::erase(std::remove_if(_customActions.begin(), _customActions.end(), isOwner), _customActions.end());
```

Reply:
I see you incorporated this change in other parts @IntelOrca , are you doing this one too?